components/director/pkg/paging/page_iterator.go

```diff
@@ -0,0 +1,89 @@
+package paging
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+//FetchPageFunc is responsible for executing an HTTP request to the url which the PageIterator provided.
+//If the results need to be further processed, they should be saved in a closure.
+//The function returns the number of the results fetched and an error if any occurred during the request execution.
+type FetchPageFunc func(string) (uint, error)
+
+//PageIterator is responsible for executing multiple HTTP requests until all results
+//of a pageable API are fetched
+type PageIterator struct {
+	baseURL               string
+	pageSkipFormat        string
+	pageSizeFormat        string
+	additionalQueryParams []string
+
+	do       FetchPageFunc
+	skip     uint
+	pageSize uint
+
+	paramsCount uint
+	nextURL     string
+}
+
+//NewPageIterator constructs a new page iterator with the given args
+func NewPageIterator(baseURL, skipFormat, sizeFormat string, additionalQueryParams []string, pageSize uint, do FetchPageFunc) PageIterator {
+	return PageIterator{
+		baseURL:               baseURL,
+		pageSkipFormat:        skipFormat,
+		pageSizeFormat:        sizeFormat,
+		additionalQueryParams: additionalQueryParams,
+		skip:                  0,
+		pageSize:              pageSize,
+		do:                    do,
+	}
+}
+
+//Next fetches the next page of the PageIterator. It returns true if there possibly are more pages that can eb fetched.
+//In order to get all the results that the PageIterator can fetch Next should be called until it returns false and no error.
+//Once next returns false with no error it should not be called anymore. If it does get called the behaviour is undefined.
+func (p *PageIterator) Next() (bool, error) {
+	p.buildNextURL()
+	count, err := p.do(p.nextURL)
+	if err != nil {
+		return false, errors.Wrapf(err, "while fetching next page: ")
+	}
+
+	if count < p.pageSize {
```

Comment:
What will happen if there are exactly `pageSize` elements returned and no more. The code using page iterator will call `Next` and it will fail to fetch any more elements. 
`p.do` function should return whether there is next page or not depending on the response.

Reply:
As discussed, the backend should respond with empty result set if there are no more elements.